src/vs/workbench/contrib/chat/browser/contrib/chatDynamicVariables.ts

```diff
@@ -60,60 +70,120 @@ export class ChatDynamicVariableModel extends Disposable implements IChatWidgetC
 							this.widget.refreshParsedInput();
 						}
 						return null;
-					} else if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
+					}
+
+					if (Range.compareRangesUsingStarts(ref.range, c.range) > 0) {
 						const delta = c.text.length - c.rangeLength;
-						return {
-							...ref,
-							range: {
-								startLineNumber: ref.range.startLineNumber,
-								startColumn: ref.range.startColumn + delta,
-								endLineNumber: ref.range.endLineNumber,
-								endColumn: ref.range.endColumn + delta
-							}
+						ref.range = {
+							startLineNumber: ref.range.startLineNumber,
+							startColumn: ref.range.startColumn + delta,
+							endLineNumber: ref.range.endLineNumber,
+							endColumn: ref.range.endColumn + delta,
 						};
+
+						return ref;
 					}
 
 					return ref;
 				}));
 			});
 
-			this.updateDecorations();
+			this.updateVariableDecorations();
 		}));
 	}
 
-	getInputState(): any {
-		return this.variables;
-	}
-
 	setInputState(s: any): void {
 		if (!Array.isArray(s)) {
 			s = [];
 		}
 
 		this._variables = s;
-		this.updateDecorations();
+		this.updateVariableDecorations();
 	}
 
-	addReference(ref: IDynamicVariable): void {
-		this._variables.push(ref);
-		this.updateDecorations();
+	public addReference(ref: IDynamicVariable): void {
+		const variable = this._register(new ChatDynamicVariable(ref, this.fileService));
+
+		this._variables.push(variable);
+		this.updateVariableDecorations();
 		this.widget.refreshParsedInput();
+
+		// if the `prompt snippets` feature is enabled, start resolving
+		// nested file references immediatelly and subscribe to updates
+		if (ChatDynamicVariable.promptSnippetsEnabled(this.configService)) {
+			// subscribe to variable changes
+			this._register(variable.onUpdate(() => {
+				this.updateVariableTexts();
+				this.updateVariableDecorations();
+				this.widget.refreshParsedInput();
+			}));
+			// make sure the variable is updated on filesystem changes
+			variable.addFilesystemListeners();
```

Comment:
I'm a bit surprised by this, usually the object manages its listeners internally to keep the public interface minimal.

Reply:
For performance reasons we want to reduce number of listeners we add - we cannot add a listener in the constructor, because we want to add it only when the feature is enabled, and the variable is for a prompt snippet file, hence we do it inside the if block here 🤷 I bet it will change in the future tho and we will be able to do this automatically as you suggest 👐

Reply:
We could check whether the feature is added inside the constructor for the object, and add the listener there?

Reply:
I had that thought but that make the class to be dependent on the `IConfigurationService` and I tried to avoid that 🤔

Reply:
Why? We inject `IConfigurationService` all the time whenever we need to check for settings enablement

Reply:
Ok updated 🙌